### ISSUE TITLE
Non deterministic behavior when iterating over paginatedList after calling search_code

### ISSUE THREAD

**Issue by alextsakpinis** [PyGithub/PyGithub#2166]
Hello community,
I am using the `search_code` function to find repositories which used a specific term in a file. When iterating over the result within a for loop and just printing the current counter to see how far the loop goes, the program does **not** reach `result.totalCount` as expected. It stops more or less randomly somewhere before. 

```
query = "\"known_third_party = \"kedro\"\" in:file language:toml"
counter = 0
result = github.search_code(query)
print(str(result.totalCount))
for code in result:
    counter += 1
    print(counter, code.repository.clone_url)
    time.sleep(1)
```

Executing the above code three times in a row, the totalCount was always 340, but the program stopped at 280, 311 and 318.

Can you please help me with that? Why does the loop not reach the last item?

Thanks a lot for your help!

With best regards,
Alex

**Comment by DSoko2** [comment]
I am facing the same issue. So far I observed two causes:

- Github API response: The response for a result page sometimes contains less items than the page should contain. E.g., if per_page is set to 30, sometimes a request for an intermediate page returns less items. The paginated list implementation ignores that fact and assumes the page is complete. Repeating the same request manually shows that sometimes the full page is returned, and sometimes subsets of different length. Strangely, even if the page is not completely returned, the incomplete_results header is false.
- If the iterator's next call is called inside some custom logic to retry on rate limit exceeded errors, a RateLimitExceededException breaks (completes?) the iterator, i.e., repeating next(iterator) after a RateLimitExceededException always raises StopIteration.

I am not sure how to solve this issues idiomatically, perhaps a maintainer can help us?

**Comment by smonsef** [comment]
I have recently faced this issue too. 
Has anyone been working on a fix?

**Comment by EnricoMi** [comment]
With `2.0.0rc1`, rate limit errors are retried by PyGithub, and thus while iterating over the paginated list. So your second point should be resolved.

Reloading a page that seems incomplete until a complete page is retrieved sounds like a bad idea, trying to fix undesired behaviour of the Github API. If a page is never completely returned, that will turn into an infinite loop.

The paginated list relies on Github API indicating the next page nor not. If a page is retrieved without an exception, it is considered complete. If there is a next page indicated, the iteration continues.

Github Support always claims that search is an expensive API to call and they try to provide best effort only. The result is not guaranteed to be complete or consistent, so that list of results returned by PyGithub should be taken with the same guarantees: as good as it gets.

**Comment by Felixoid** [comment]
> Reloading a page that seems incomplete until a complete page is retrieved sounds like a bad idea, trying to fix undesired behaviour of the Github API. If a page is never completely returned, that will turn into an infinite loop.

That's true. In our case, we solved the issue by adding another filter with shorter data range to the query additional to a very wide one. Wide `merged:X..Y` plus narrower `updated:A..B`

> The result is not guaranteed to be complete or consistent, so that list of results returned by PyGithub should be taken with the same guarantees: as good as it gets.

That's as well true, but having the `incomplete_results` property is necessary to understand if the result at least could be trusted or not. Currently, this flag is just discarded.
